### PR TITLE
DL: Add support for reporting various metrics in fit/evaluate

### PR DESCRIPTION
**Pull request by jingyimei**
This PR aims to improve the way metrics are handled. 
1. We add support for various metrics
2. We validate to make sure only a single metric is passed
3. We add support for metrics_compute_frequency for training data
4. We remove metric and loss computation from fit transition

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/915/

File: src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in [hunk 1]
```diff
@@ -39,9 +40,9 @@ def get_model_shapes(model):
 def deserialize_weights(model_state, model_shapes):
     """
     Parameters:
-        model_state: a stringified (serialized) state containing loss,
-        accuracy, image_count, and model_weights, passed from postgres
-        model_shapes: a list of tuples containing the shapes of each element
+        model_state: a stringified (serialized) state containing
+        image_count and model_weights, passed from postgres model_shapes:
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in]: new line after postgres

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 2]
```diff
@@ -282,30 +270,30 @@ def fit(schema_madlib, source_table, model,model_arch_table,
...
             $4 AS description,
             '{model_type}'::TEXT AS model_type,
-            {model_size} AS model_size,
+            {model_size}::INTEGER AS model_size,
             '{start_training_time}'::TIMESTAMP AS start_training_time,
             '{end_training_time}'::TIMESTAMP AS end_training_time,
             $5 AS time_iter,
             '{version}'::TEXT AS madlib_version,
-            {num_classes} AS num_classes,
+            {num_classes}::INTEGER AS num_classes,
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: why do we need to typecast this one and the others to integer ? Aren't they integer by default ?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 3]
```diff
@@ -342,23 +330,24 @@ def compute_loss_and_metrics(schema_madlib, table, dependent_varname,
...
+                                                      seg_ids_val,
+                                                      rows_per_seg_val,
+                                                      gp_segment_id_col)
     end_val = time.time()
     plpy.info("Time for evaluation in iteration {0}: {1} sec.". format(
         curr_iter, end_val - start_val))
-    if len(evaluate_result) < 2:
-        plpy.error('Calling evaluate on table {0} returned < 2 '
-                   'metrics. Expected both loss and a metric.'.format(
+    if len(evaluate_result) not in [1, 2]:
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: Since metric can be an optional param, isn't it possible that keras.eval only returns the loss ?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 4]
```diff
@@ -459,14 +448,10 @@ def fit_transition(state, ind_var, dep_var, current_seg_id, num_classes,
...
         image_count = 0
-        agg_loss = 0
-        agg_accuracy = 0
         agg_image_count = 0
     else:
         segment_model = SD['segment_model']
-        #TODO we don't need to deserialize the weights here.
-        agg_loss, agg_accuracy, agg_image_count, _ = madlib_keras_serializer.deserialize_weights(
-            state, SD['model_shapes'])
+        agg_image_count = float(np.fromstring(state, dtype=np.float32)[0])
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: It's good that we are not deserializing the weights just to get the image count but can we move this code to the serializer file so that all the logic stays in one place?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 5]
```diff
@@ -638,21 +611,21 @@ def internal_keras_eval_transition(state, dependent_var, independent_var,
...
-        _, _, _, model_weights = madlib_keras_serializer.deserialize_weights(
+        _, model_weights = madlib_keras_serializer.deserialize_weights(
             model_data, model_shapes)
         model.set_weights(model_weights)
         with K.tf.device(device_name):
             compile_model(model, compile_params)
         SD['segment_model'] = model
         # These should already be 0, but just in case make sure
-        agg_accuracy = 0
+        agg_metric = 0
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: Calling madlib_keras_fit without any metrics fails with the following error

```
SELECT madlib.madlib_keras_fit('cifar_10_sample_batched','keras_saved_out','model_arch',1, 
$$ optimizer=SGD(lr=0.01, decay=1e-6, nesterov=True), loss='categorical_crossentropy'$$::text, 
$$ batch_size=2, epochs=1, verbose=0 $$::text,3);

ERROR:  plpy.SPIError: TypeError: 'numpy.float64' object is not iterable (plpython.c:5038)  (seg0 slice1 127.0.0.1:25432 pid=95971) (plpython.c:5038)
DETAIL:
Traceback (most recent call last):
  PL/Python function "internal_keras_eval_transition", line 6, in <module>
    return madlib_keras.internal_keras_eval_transition(**globals())
  PL/Python function "internal_keras_eval_transition", line 639, in internal_keras_eval_transition
PL/Python function "internal_keras_eval_transition"
CONTEXT:  Traceback (most recent call last):
  PL/Python function "madlib_keras_fit", line 21, in <module>
    madlib_keras.fit(**globals())
  PL/Python function "madlib_keras_fit", line 216, in fit
  PL/Python function "madlib_keras_fit", line 343, in compute_loss_and_metrics
  PL/Python function "madlib_keras_fit", line 602, in get_loss_metric_from_keras_eval
PL/Python function "madlib_keras_fit"
```

We should add a dev-check test for this use case and also for when metrics is passed as an empty string ` metrics=['']`

File: src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in [hunk 6]
```diff
@@ -18,6 +18,7 @@
 # under the License.
 
 import numpy as np
+import plpy
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in]: we don't need this plpy import do we ?

File: src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in [hunk 7]
```diff
@@ -110,12 +110,16 @@ SELECT madlib_keras_fit(
...
     'model_arch',
     1,
-    $$ optimizer=SGD(lr=0.01, decay=1e-6, nesterov=True), loss='categorical_crossentropy', metrics=['accuracy']$$::text,
+    $$ optimizer=SGD(lr=0.01, decay=1e-6, nesterov=True), loss='categorical_crossentropy', metrics=['mae']$$::text,
     $$ batch_size=2, epochs=1, verbose=0 $$::text,
     3,
     NULL,
     'cifar_10_sample_val');
 
+\x
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in]: Do we need this select statement ?

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 2 above)
Comment (jingyimei) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: If we don't type cast, it will be an unknown type in those column, and postgres will give a warning, so that the whole create summary table sql will be printed out in the log.

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 3 above)
Comment (jingyimei) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: Addressed

File: src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in (hunk 1 above)
Comment (jingyimei) [reply src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in]: addressed

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 4 above)
Comment (jingyimei) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: Addressed

File: src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in (hunk 6 above)
Comment (jingyimei) [reply src/ports/postgres/modules/deep_learning/madlib_keras_serializer.py_in]: Removed

File: src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in (hunk 7 above)
Comment (jingyimei) [reply src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in]: removed

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/918/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/919/

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 5 above)
Comment (jingyimei) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: metrics is not specified and metrics=[] are addressed, so that they won't be passed as part of compile params. But for case metrics=[''], keras will throw out error saying unknown metric. I would let keras handle this, instead of us. Thoughts? @kaknikhil

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in [hunk 8]
```diff
@@ -637,13 +642,21 @@ def internal_keras_eval_transition(state, dependent_var, independent_var,
...
+    else:
+        loss = res
+        metric = None
 
     image_count = len(dependent_var)
 
     agg_image_count += image_count
     agg_loss += (image_count * loss)
-    agg_metric += (image_count * metric)
+    plpy.info("agg_metric is {}".format(agg_metric))
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: again do we need this information to be printed to the console ? there are a few more plpy infos that need to be removed.

File: src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in [hunk 9]
```diff
@@ -270,6 +270,17 @@ FROM (SELECT * FROM keras_out_summary) summary;
...
+SELECT madlib.madlib_keras_fit(
+'cifar_10_sample_batched',
+'keras_saved_out',
+'model_arch',
+1,
+$$ optimizer=SGD(lr=0.01, decay=1e-6, nesterov=True), loss='categorical_crossentropy'$$::text,
+$$ batch_size=2, epochs=1, verbose=0 $$::text,
+3);
+
 DROP TABLE IF EXISTS cifar10_predict;
```
Comment (kaknikhil) [inline src/ports/postgres/modules/deep_learning/test/madlib_keras.sql_in]: maybe add a test for ` $$ optimizer=SGD(lr=0.01, decay=1e-6, nesterov=True), loss='categorical_crossentropy, metrics=[]'$$::text,`

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 5 above)
Comment (kaknikhil) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: yeah i think it's fine to let keras fail for `metrics=['']`

**Comment by kaknikhil** [comment]
What do we output in the metric column of the summary table if metric is not passed in ? Is it 0 or null ?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/923/

File: src/ports/postgres/modules/deep_learning/madlib_keras.py_in (hunk 3 above)
Comment (jingyimei) [reply src/ports/postgres/modules/deep_learning/madlib_keras.py_in]: evaluate_result is always a list, and evaluate_result[0] is always loss, and we only allow one more metric here, so that the valid length of evaluate_result should either 1 or 2

**Comment by jingyimei** [comment]
> What do we output in the metric column of the summary table if metric is not passed in ? Is it 0 or null ?

@kaknikhil for now, we leave it as 0, we will take care of it in a future PR.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/924/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/925/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/madlib-pr-build/927/
